src/devices/machine/i8291a.h

```diff
@@ -0,0 +1,417 @@
+	static constexpr uint8_t IFCMD_GTL        = 0x01;  // Go to local
+	static constexpr uint8_t IFCMD_SDC        = 0x04;  // Selected device clear
+	static constexpr uint8_t IFCMD_GET        = 0x08;  // Group execute trigger
+	static constexpr uint8_t IFCMD_TCT        = 0x09;  // Take control
+	static constexpr uint8_t IFCMD_LLO        = 0x11;  // Local lock-out
+	static constexpr uint8_t IFCMD_DCL        = 0x14;  // Device clear
+	static constexpr uint8_t IFCMD_SPE        = 0x18;  // Serial poll enable
+	static constexpr uint8_t IFCMD_SPD        = 0x19;  // Serial poll disable
+	static constexpr uint8_t IFCMD_UNL        = 0x3f;  // Unlisten
+	static constexpr uint8_t IFCMD_UNT        = 0x5f;  // Untalk
+
+	enum class source_handshake_state {
+		SIDS,
+		SGNS,
+		SDYS,
+		STRS
+	};
+
+	enum class acceptor_handshake_state {
+		AIDS,
+		ANRS,
+		ACRS,
+		AWNS,
+		ADYS,
+		ACDS
+	};
+
+	enum class talker_state {
+		TIDS,
+		TADS,
+		SPAS,
+		TACS
+	};
+
+	enum class talker_primary_state {
+		TPIS,
+		TPAS
+	};
+
+	enum class listener_primary_state {
+		LPIS,
+		LPAS
+	};
+
+	enum class talker_serial_poll_state {
+		SPIS,
+		SPMS
+	};
+
+	enum class serial_poll_state {
+		NPRS,
+		SRQS,
+		APRS
+	};
+
+	enum class listener_state {
+		LIDS,
+		LADS,
+		LACS
+	};
+
+	enum class remote_local_state {
+		LOCS,
+		REMS,
+		RWLS,
+		LWLS
+	};
+
+	enum class parallel_poll_state {
+		PPIS,
+		PPSS,
+		PPAS
+	};
+
+	enum class device_clear_state {
+		DCIS,
+		DCAS
+	};
+
+	enum class device_trigger_state {
+		DTIS,
+		DTAS
+	};
+
+	static const char *get_state_name(acceptor_handshake_state state);
+	static const char *get_state_name(source_handshake_state state);
+	static const char *get_state_name(talker_state state);
+	static const char *get_state_name(talker_primary_state state);
+	static const char *get_state_name(talker_serial_poll_state state);
+	static const char *get_state_name(listener_state state);
+	static const char *get_state_name(listener_primary_state state);
+	static const char *get_state_name(device_clear_state state);
+	static const char *get_state_name(device_trigger_state state);
+	static const char *get_state_name(parallel_poll_state state);
+	static const char *get_state_name(serial_poll_state state);
+	static const char *get_state_name(remote_local_state state);
+
+	template<typename T> void update_state(T &name, T state, int timeout = 0)
+	{
+			if (name != state) {
+				m_state_changed = true;
+				//logerror("%s: %s -> %s\n", __FUNCTION__,
+				//		get_state_name(name),
+				//		get_state_name(state));
+				name = state;
+			}
+	}
+
+	source_handshake_state m_sh_state;
+	acceptor_handshake_state m_ah_state;
+	talker_state m_t_state;
+	talker_primary_state m_tp_state;
+	talker_serial_poll_state m_tsp_state;
+	listener_state m_l_state;
+	listener_primary_state m_lp_state;
+	remote_local_state m_rl_state;
+	serial_poll_state m_sp_state;
+	parallel_poll_state m_pp_state;
+	device_clear_state m_dc_state;
+	device_trigger_state m_dt_state;
```

Comment:
You're going to have to save these in save state as well.  To make them savable, use the `ALLOW_SAVE_TYPE` macro, and specify an explicit storage class for each enum (see examples in src/devices/cpu/mcs40/mcs40.h and src/devices/cpu/mcs40/mcs40.cpp).

Reply:
fixed